### PR TITLE
testing: LocalChangeUtilsTest Implemented in dao core module

### LINKED ISSUE #187: Tests for LocalChangeUtils
> did you add tests for this static class? if not, please create an issue as a follow up :)

i think one of the benefits of having this static class is that the tests can be written very independently.

_Originally posted by @jingtang10 in https://github.com/google/android-fhir/pull/146#r564426060_

### PR DESCRIPTION
**Pull request by Ana2k**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #187 

**Description**
Implemented Unit Test for LocalChangeUtils. Checked using dummy JSON response, whether INSERT, UPDATE, DELETE and their combination, returned apt values for the dao file function.

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?
N/A

**Type**
Choose one: Testing

**Screenshots (if applicable)**
N/A

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md)
- [x] I have read [How to Contribute](https://github.com/google/android-fhir/blob/master/docs/contributing.md)
- [x] I have read the [Developer's guide](https://github.com/google/android-fhir/wiki/Developer's-Guide)
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate )
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotless check to check my code follows the style guide of this project
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally
- [x] I have built and run the reference app(s) to verify my change fixes the issue and/or does not break the reference app(s)

### CODE REVIEW

**Comment by Ana2k** [comment]
@deepankarb @jingtang10 I have implemented the tests. Please review this pr.
Thanks!

File: core/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+                    "use": "usual",
+                     "given": [
+                     "Kenzi"
+                    ]
+                    }
+           ]}"""
+
+    val json_patch =
+      """[{"op": "replace",
+                    |"path": "/name/0/use",
```
Comment (jingtang10) [inline core/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:51]: Do you mind removing the `|`s? do they affect your result at all?

File: core/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+          "resourceType": "Patient",
+          "id": "human",
+          "name": [
+                  {
+                    "use": "usual",
+                     "given": [
+                     "Kenzi"
+                    ]
+                    }
+           ]}"""
```
Comment (jingtang10) [inline core/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:47]: I would probably do the indentation this way:

```suggestion
      """
      {
        "resourceType": "Patient",
        "id": "human",
        "name": [
          {
            "use": "usual",
            "given": [
              "Kenzi"
            ]
          }
        ]
      }
      """
```

File: core/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      LocalChangeEntity(
+        id = 0L,
+        resourceType = "Test4",
+        resourceId = "Anam4",
+        timestamp = "",
+        type = LocalChangeEntity.Type.INSERT,
+        payload =
+          """{"resourceType":"Patient","id":"human","name":[{"use":"Ana2k","given":["Kenzi"]}]}"""
+      )
+    assertEquals(mergeLocalChanges(fifth, fourth), res54)
```
Comment (jingtang10) [inline core/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:151]: Please separate these into 4 different test cases each with a descriptive name

```
@Test
mergeInsertAndInsert

@Test
mergeInsertAndDelete

...
...

```


The reason to do things this way is that each test will be independent. So if a test is failing you can tell by looking at the test case name to figure out what is happening. If your test case fails now you will have to check the stack trace or even debug to figure out which case is actually failing.

**Review by jingtang10 — CHANGES_REQUESTED** [review]
Thanks @Ana2k!

File: core/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt (hunk 3 above)
Comment (Ana2k) [reply core/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:151]: > Please separate these into 4 different test cases each with a descriptive name
> 
> ```
> @Test
> mergeInsertAndInsert
> 
> @Test
> mergeInsertAndDelete
> 
> ...
> ...
> ```
> 
> The reason to do things this way is that each test will be independent. So if a test is failing you can tell by looking at the test case name to figure out what is happening. If your test case fails now you will have to check the stack trace or even debug to figure out which case is actually failing.

Thanks for the clear explanation. Will get these done in a couple of days :)

**Comment by jingtang10** [comment]
please re-request review when this is done thanks

**Comment by Ana2k** [comment]
@jingtang10 I have implemented the required changes please review. Thanks!

**Comment by Ana2k** [comment]
@jingtang10 @vincent-karuri please review this pr. Thanks.

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+        timestamp = "",
+        type = LocalChangeEntity.Type.DELETE,
+        payload = ""
+      )
+    assertEquals(mergeLocalChanges(first, third), res13)
+  }
+
+  @Test
+  fun mergeDeleteAndInsert() {
+    // Expected output when third and first are mixed( DELETE AND INSERT)
```
Comment (vincent-karuri) [inline engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:142]: I think we can move the comments to form part of the test name i.e. make the test name more descriptive to specify the SUT, condition, and expected behavior. e.g. https://github.com/google/android-fhir/blob/4892adbddf78f4fdef40d3f95bd5e2bb19424fdf/engine/src/androidTest/java/com/google/android/fhir/db/impl/DatabaseImplTest.kt#L119

reference: https://source.android.com/setup/contribute/code-style#javatests-style-rules

Comment (vincent-karuri) [reply engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:142]: we can also rename the resources to describe they're content better. atm, it's not too clear what the numbers mean i.e. why `res31, res54 etc.`. Unless it's not important to distinguish between them.

Comment (jingtang10) [reply engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:142]: hi vincent - do you mind adding this reference: https://source.android.com/setup/contribute/code-style#javatests-style-rules to the developer's guide? https://github.com/google/android-fhir/wiki/Developer's-Guide in the naming convention section. thanks!

Comment (Ana2k) [reply engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt]: > we can also rename the resources to describe their content better. atm, it's not too clear what the numbers mean i.e. why `res31, res54 etc.`. Unless it's not important to distinguish between them.

@vincent-karuri @jingtang10 
I have made the required changes I as well as I could. 
Please review this pr for further changes. Thanks!

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    )
+  val insert_payload_json_node =
+    LocalChangeEntity(
+      id = 5L,
+      resourceType = "Test5",
+      resourceId = "Anam5",
+      timestamp = Date().toTimeZoneString(),
+      payload = payload_json_node.toString(),
+      type = LocalChangeEntity.Type.INSERT
+    )
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:108]: Please inline all these variables - I would much rather have some duplication in each test cases but do not need to scroll up and down to figure out what the behaviour needs to be.

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+  @Test
+  fun response_test_insert_payload1_insert_payload2() {
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:111]:
```suggestion
  fun mergeLocalChanges_insertTwice() {
```

same with the other names

**Review by jingtang10 — CHANGES_REQUESTED** [review]

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt (hunk 6 above)
Comment (Ana2k) [reply engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt]: @jingtang10  @vincent-karuri I have implemented all the required changes. Kindly review this pr. Thanks!

**Comment by Ana2k** [comment]
@jingtang10 @vincent-karuri I have implemented all the required changes. Kindly review this pr. Thanks!

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 7]
```diff
@@ -0,0 +1,206 @@
...
+     """.trimMargin()
+
+  // converting the json payload to json node type
+  val objectMapper = ObjectMapper()
+  val payload_json_node: JsonNode = objectMapper.readTree(payload1)
+
+  @Test
+  fun mergeLocalChnages_insertTwice() {
+    // Expected output when insert_payload1 and insert_payload2 mix
+    // (INSERT AND INSERT)
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:67]: remove this comment which is not adding anything

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+          }
+       ]
+     """.trimMargin()
+
+  // converting the json payload to json node type
+  val objectMapper = ObjectMapper()
+  val payload_json_node: JsonNode = objectMapper.readTree(payload1)
+
+  @Test
+  fun mergeLocalChnages_insertTwice() {
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:65]:
```suggestion
  fun mergeLocalChanges_insert_insert() {
```

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+        timestamp = "",
+        type = LocalChangeEntity.Type.DELETE,
+        payload = ""
+      )
+    assertThat(mergeLocalChanges(insert_payload1, delete_payload))
+      .isEqualTo(insertPayload1_and_deletePayload)
+  }
+
+  @Test
+  fun mergeLocalChanges_delete_and_insert() {
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:136]:
```suggestion
  fun mergeLocalChanges_delete_insert() {
```

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 10]
```diff
@@ -0,0 +1,206 @@
+/*
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:1]: Can you modify `LocalChangeUtils.mergeLocalChanges` and make sure the rresource type and resource ids match?
```
    check(first.resourceType == second.resourceType) {
      "Resource types ${first.resourceType} and ${second.resourceType} do not match"
    }
    check(first.resourceId == second.resourceId) {
      "Resource IDs ${first.resourceId} and ${second.resourceId} do not match"
    }
```

And then add 2 test cases: 1 with different resource types and 1 with different resource ids and assert that exceptions are thrown.

look for the use of `assertThrows` in the codebase

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 11]
```diff
@@ -0,0 +1,206 @@
...
+
+    val insertPayload1_and_insertPayload2 =
+      LocalChangeEntity(
+        id = 0L,
+        resourceType = "Test2",
+        resourceId = "Anam2",
+        timestamp = "",
+        type = LocalChangeEntity.Type.INSERT,
+        payload = "payload2"
+      )
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:95]: inline these variables?

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 12]
```diff
@@ -0,0 +1,206 @@
...
+          "op": "replace",
+          "path": "/name/0/use",
+          "value":"Ana2k"          
+          }
+       ]
+     """.trimMargin()
+
+  // converting the json payload to json node type
+  val objectMapper = ObjectMapper()
+  val payload_json_node: JsonNode = objectMapper.readTree(payload1)
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:62]: no need to be class variables if only used in the last test case?

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 13]
```diff
@@ -0,0 +1,206 @@
...
+import com.google.android.fhir.toTimeZoneString
+import com.google.common.truth.Truth.assertThat
+import java.util.Date
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.RobolectricTestRunner
+
+/** Unit tests for [LocalChangeUtils]. */
+@RunWith(RobolectricTestRunner::class)
+class LocalChangeUtilsTest {
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:32]: overall you need to have 3^2 = 9 test cases:

insert insert
insert update
insert delete
update insert
update update
update delete
delete insert
delete update
delete delete

here you've got 5 so 4 are still missing

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+        timestamp = "",
+        type = LocalChangeEntity.Type.INSERT,
+        payload = "payload1"
+      )
+    assertThat(mergeLocalChanges(delete_payload, insert_payload1))
+      .isEqualTo(deletePayload_and_insertPayload1)
+  }
+
+  @Test
+  fun mergeLocalChanges_insertPayloadJNode_and_UpdateJpatch() {
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:172]:
```suggestion
  fun mergeLocalChanges_insert_update() {
```

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+    // Expected output when insert_payload_json_node and update_json_patch are mixed
+    // (INSERT-PayloadJNODE AND UPDATE-JSON-PATCH)
```
Comment (jingtang10) [inline engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:174]: remove this comment - it's not adding anything or explaining anything more than the test case name

**Review by jingtang10 — CHANGES_REQUESTED** [review]
Thanks @Ana2k !

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt (hunk 11 above)
Comment (Ana2k) [reply engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:95]: I have inlined these for now, but its too clumpsy and non readable, according to me.
I saw the format at https://github.com/google/android-fhir/blob/1c3d04de5ba90d209c539be514a571ec371ca9ed/engine/src/main/java/com/google/android/fhir/db/impl/dao/LocalChangeUtils.kt#L65
that's why stuck to the same.

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt (hunk 10 above)
Comment (Ana2k) [reply engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:1]: @jingtang, as per my understanding of the codebase, LocalChangeUtils.mergeLocalChanges support merging of  different resourceType ,as shown in the if else statements given

Bu defenintely I can add check for matching of IDs. So implementing that as of now. Please tell your views

**Comment by Ana2k** [comment]
@jingtang10 I have implemented all the required changes.
Except the resourceType spec 
Reason in :https://github.com/google/android-fhir/pull/424#discussion_r649335620

Kindly please review this pr. Thanks!

**Comment by vincent-karuri** [comment]
@jingtang10 I think we have the same integration test issue here as well

File: engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt [hunk 16]
```diff
@@ -0,0 +1,468 @@
...
+    val exception: Throwable =
+      Assert.assertThrows(IllegalArgumentException::class.java) {
+        mergeLocalChanges(delete_payload, update_json_patch)
+      }
+    assertThat("Cannot merge local changes with type DELETE and UPDATE.")
+      .isEqualTo(exception.message)
+  }
+
+  @Test
+  fun mergeLocalChanges_delete_delete() { // delete delete
```
Comment (vincent-karuri) [inline engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:435]: hi @Ana2k were you able to take a look at the format on the other tests https://github.com/google/android-fhir/pull/424#discussion_r640355991?

Comment (vincent-karuri) [reply engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt:435]: It follows the format recommended here https://github.com/google/android-fhir/pull/424#discussion_r640355991 with a slight modification.

Basically, this test method should read something like `mergeLocalChanges_deletion_shouldMergeDeletePayloads`.  This allows one to quickly understand all the "variables" at play in the test by simply looking at the test name. 

Also, it removes the need for comments in the test as the name is self-explanatory.

Comment (Ana2k) [reply engine/src/test/java/com/google/android/fhir/db/impl/dao/LocalChangeUtilsTest.kt]: > hi @Ana2k were you able to take a look at the format on the other tests [#424 (comment)](https://github.com/google/android-fhir/pull/424#discussion_r640355991)?

@vincent-karuri  Thanks for the specifics.Theye were really helpful.I have made relevant changes to the best of my understanding.
Can @jingtang10 and you review this pr and mention if more changes are required?
Thanks!

**Comment by Ana2k** [comment]
@jingtang10 reminder for reviewing this pr. 
@maimoonak tagging you to update about previous PRs.
Thanks!

**Comment by Tarun-Bhardwaj** [comment]
Hi @Ana2k , is there an ETA you can provide for completing the changes as per the review comments in order to close this PR?

**Comment by Ana2k** [comment]
> Hi @Ana2k , is there an ETA you can provide for completing the changes as per the review comments in order to close this PR?

Hi @Tarun-Bhardwaj , almost all of the required changes have been implemented for this. ETA is 10th August for this.

**Comment by jingtang10** [comment]
Closing this since it's not been open worked on.

**Pull request closed without merging**